File: 287-find-the-duplicate-number/287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int n = nums.length;
//          T.C = 0(N^2) .. S.C = 0(1)
        
        // for(int i=0;i<n-1;i++){
        //     for(int j=i+1;j<n;j++){
        //         if(nums[i]==nums[j]){
        //             return nums[i];
        //         }
        //     }
        // }
        // return 1;
        
        HashMap<Integer,Integer> map = new HashMap<>();
        for(int i=0;i<n;i++){
            if(map.containsKey(nums[i])){
                return nums[i];
            }
            
            map.put(nums[i],1);
        }
        return 1;
    }
}

File: 287-find-the-duplicate-number/287-find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
//         brute force time complexity 0(n) .. space complexity 0(n)
        
        
       // unordered_map<int,int> umap;
       //  for(auto &i:nums){
       //      if(umap.find(i)!=umap.end())return i;
       //      else umap[i]++;
       //  }
       //  return 1;
        
//          better solution  time complexity 0(n) space complexity 0(1)
        
        // for(int i=0;i<nums.size();i++){
        //     for(int j=0;j<nums.size();j++){
        //         if(i!=j && nums[i]==nums[j]){
        //             return nums[i];
        //         }
        //     }
        // }
        // return 1;
        
        for(int i=0;i<nums.size();i++){
            int index = abs(nums[i])-1;
            nums[index]*=-1;
            if(nums[index]>0) return abs(nums[i]);
        }
        return -1;
    }
};
 